components/style/dom.rs

```diff
@@ -64,7 +64,33 @@ pub trait TRestyleDamage : Debug + PartialEq + BitOr<Output=Self> + Copy {
     fn rebuild_and_reflow() -> Self;
 }
 
-pub trait TNode : Sized + Copy + Clone {
+/// Simple trait to provide basic information about the type of an element.
+///
+/// We avoid exposing the full type id, since computing it in the general case
+/// would be difficult for Gecko nodes.
+pub trait NodeInfo {
+    fn is_element(&self) -> bool;
+    fn is_text_node(&self) -> bool;
+
+    // Comments, doctypes, etc are ignored by layout algorithms.
+    fn needs_layout(&self) -> bool { self.is_element() || self.is_text_node() }
+}
+
+pub struct LayoutIterator<T>(pub T);
+impl<T, I> Iterator for LayoutIterator<T> where T: Iterator<Item=I>, I: NodeInfo {
+    type Item = I;
+    fn next(&mut self) -> Option<I> {
+        loop {
+            // Filter out nodes that layout should ignore.
+            let n = self.0.next();
```

Comment:
nit: maybe this would look clearer as:

```rust
match self.0.next() {
    None => return None,
    Some(node) if node.needs_layout() => return Some(node),
    _ => {},
}
```

Though I don't have a strong preference, so feel free to leave as-is.